rfcs/0005-range-queries.md

```diff
@@ -0,0 +1,258 @@
+* https://github.com/facebook/rocksdb/wiki/Iterator
+* https://github.com/facebook/rocksdb/wiki/Iterator-Implementation
+* https://github.com/slatedb/slatedb/pull/260
+* https://github.com/slatedb/slatedb/blob/main/rfcs/0004-checkpoints.md
+
+## Motivation
+
+
+The ability to scan the keys in a range is an essential feature for building applications.
+For example, this can be used to implement a collections of keys using a common prefix. 
+Currently, SlateDB only supports fetching individual values for known keys.
+
+## Goals
+
+The goal of this work is to provide a flexible capability to query a range of keys in the
+database. It should be possible to specify unbounded ranges to allow querying all keys,
+or bounded ranges for finer precision and efficiency.
+
+Range queries have the potential to be expensive from a resource perspective because they
+hold references to existing tables, which prevents garbage collection. Our goal here is
+to provide a simple initial approach to state management which optimizes for common cases,
+while leaving room for further optimization in the future.
+
+## Public API
+
+We introduce a new `Db::scan` API which accept a `DbRange` defining the range of the scan.
+The result will be a `DbIterator` object which iterates the current records within the range.
+Users may either provide an explicit starting range in the call to `scan`, or they
+may `seek` using the returned iterator.
+
+```rust
+impl Db {
+    /// Scan a range of keys.
+    /// 
+    /// returns a `DbIterator`
+    pub async fn scan<T>(
+        &self,
+        range: T
+    ) -> Result<DbIterator, SlateDbError> where
+        T: RangeBounds<Bytes> 
+    {
+	    ...
+    }
+	
+}
+
+impl DbIterator {
+    /// Get the next record in the scan.
+    /// 
+    /// returns Ok(None) when the scan is complete
+    /// returns Err(InvalidatedIterator) if the iterator has been invalidated
+    ///  due to an underlying error
+    pub async fn next(
+        &mut self,
+    ) -> Result<Option(DbRecord), SlateDbError> {
+        ...
+    }
+
+    /// Seek to a key ahead of the last key returned from the iterator or
+    /// the lower range bound if no records have yet been returned. 
+    /// 
+    /// returns Ok(()) if the position is successfully advanced
+    /// returns SlateDbError::InvalidArgument if `lower_bound` is `Unbounded`
+    /// returns SlateDbError::InvalidArgument if the key is comes before the
+    ///  current iterator position
+    /// returns SlateDbError::InvalidArgument if `lower_bound` is beyond the
+    ///  upper bound specified in the original `scan` parameters
+    /// returns Err(InvalidatedIterator) if the iterator has been invalidated
+    ///  in order to reclaim resources
+    pub async fn seek(
+        &mut self,
+        lower_bound: Bound<&[u8]>
+    ) -> Result<(), SlateDbError> {
+        ...
+    }	
+
+    /// Get the id of the checkpoint that this scan references.
+    pub fn checkpoint(&self) -> UUID {
```

Comment:
what's the reason for exposing the checkpoint in the public api? It's an implementation detail, so unless it's really useful for the user, it seems better to hide it in case we want to change the implementation later.

Reply:
+1, I think checkpoint() could be made private first, as it would expose fewer details in the interface

Reply:
It was a suggestion from the comment here: https://github.com/slatedb/slatedb/pull/280#discussion_r1817952971. The use case is around persistent queries. I didn't see much harm since checkpoint IDs are already exposed to users, but I think your point is still fair. I'll take it out for now and we can consider it in a separate proposal.